rust/connlib/tunnel/src/lib.rs

```diff
@@ -45,13 +45,19 @@ const FIREZONE_MARK: u32 = 0xfd002021;
 pub type GatewayTunnel<CB> = Tunnel<CB, GatewayState>;
 pub type ClientTunnel<CB> = Tunnel<CB, ClientState>;
 
-/// Tunnel is a wireguard state machine that uses webrtc's ICE channels instead of UDP sockets to communicate between peers.
+/// [`Tunnel`] glues together connlib's [`Io`] component and the respective (pure) state of a client or gateway.
+///
+/// Most of connlib's functionality is implemented as a pure state machine in [`ClientState`] and [`GatewayState`].
+/// The only job of [`Tunnel`] is to take input from the TUN [`Device`], [`Sockets`] or time and pass it to the respective state.
```

Comment:
🤔  And the Portal connection via WebSocket / phoenix channel is not part of that pure state machine, right? It's also part of the Tunnel?

So for my goal of exercising the platform-specific I/O code of the Clients, even in Windows where a Portal can't be run, I could make a mock Tunnel implementation that uses the pure state machine but has fake I/O, similar to your prop tests, and then has a simulated portal?

Or is the WebSocket managed even higher, in `ClientTunnel` and `GatewayTunnel` wrappers?

Reply:
I think it _should_ be part of the `Tunnel`. There is no reason why the `Eventloop`s for gateways and client couldn't be in `Tunnel` (those manage the portal connection). Doing that is tracked in https://github.com/firezone/firezone/issues/3837.

But even once that is done, the connection to the portal is a side-effect and thus would likely be contained in `Io` or live at least next to it.

`ClientState` and `GatewayState` are already pure state machines so you can already run them on Windows without a portal or TUN device.

> So for my goal of exercising the platform-specific I/O code of the Clients, even in Windows

Our proptests mock everything. IO, portal, TUN device. Our (linux) integration tests don't mock anything, they use a real portal, real TUN device and real sockets.

I don't think it makes much sense to mock only parts of this. If we want to test sockets and TUN device on Windows, and thus perform real network communication, we already have an abstraction layer that can cross OS boundaries: the network. I'd try to somehow connect to a real portal and real gateway from a Windows-VM. Like, can we run a Linux container in a Windows VM on GitHub actions and connect to that?

Reply:
Actually, maybe we could. Github docs say the Windows runners have Docker installed.

But I haven't tried it because I don't want to spend all day fighting the CI again. Maybe if it was done in a side project so every run doesn't take 6 CPU-hours 🤔 And then integrated if it succeeds

It's just that, that's one more thing that has to be downloaded, installed, set up, another few gigabytes of junk to update. Testing everything end-to-end is great but I'd also just like to test the Windows to Windows Client integration sometimes.

Reply:
> It's just that, that's one more thing that has to be downloaded, installed, set up, another few gigabytes of junk to update. Testing everything end-to-end is great but I'd also just like to test the Windows to Windows Client integration sometimes.

Yeah that is definitely worthwhile testing. I've found this: https://github.com/actions/runner-images/issues/2216#issuecomment-740429928

Seems like we need a self-hosted Windows runner for that.